packages/flutter/lib/src/painting/image_resolution.dart

```diff
@@ -328,27 +327,23 @@ class AssetImage extends AssetBundleImageProvider {
     return completer.future;
   }
 
-  /// Parses the asset manifest string into a strongly-typed map.
+  /// Decodes the asset manifest's file contents into it's Dart representation.
   @visibleForTesting
-  static Future<Map<String, List<String>>?> manifestParser(String? jsonData) {
-    if (jsonData == null) {
-      return SynchronousFuture<Map<String, List<String>>?>(null);
-    }
-    // TODO(ianh): JSON decoding really shouldn't be on the main thread.
-    final Map<String, dynamic> parsedJson = json.decode(jsonData) as Map<String, dynamic>;
-    final Iterable<String> keys = parsedJson.keys;
-    final Map<String, List<String>> parsedManifest = <String, List<String>> {
-      for (final String key in keys) key: List<String>.from(parsedJson[key] as List<dynamic>),
+  static Future<Map<String, List<String>>> decodeAssetManifest(ByteData data) {
+    final Object? decoded = const StandardMessageCodec().decodeMessage(data);
+    final Map<dynamic, dynamic> typed = decoded! as Map<dynamic, dynamic>;
+    final Map<String, List<String>> result = <String, List<String>>{
+      for (final MapEntry<dynamic, dynamic> entry in typed.entries)
+        entry.key as String: (entry.value as List<dynamic>).cast<String>(),
     };
-    // TODO(ianh): convert that data structure to the right types.
-    return SynchronousFuture<Map<String, List<String>>?>(parsedManifest);
+    return SynchronousFuture<Map<String, List<String>>>(result);
```

Comment:
Nothing in this API is async, I would remove the Future usage and just return the object. Also, if you can avoid the casting, that would speed this up a bit. We're already breaking the API, so you can leave the new one as untyped.

Reply:
> Nothing in this API is async, I would remove the Future usage and just return the object.

Since `loadStructuredData` takes an asynchronous `parser`, I want to preserve that property in `loadStructuredDataBinary` to make it easier for customers to transition to. In my latest commit, I changed the type of the `parser` parameter to be a function that returns a `FutureOr<T>`. Let me know if this isn't idiomatic use of `FutureOr`--if so, we could introduce a `-Sync` version of the method.

> Also, if you can avoid the casting, that would speed this up a bit. We're already breaking the API, so you can leave the new one as untyped.

Okay, I removed all the `List.cast` calls since they definitely add up for large manifests. `decodeAssetManifest` seems to be twice as fast now. I left in the cast to `Map<dynamic, dynamic>` since we'll need to cast anyway to avoid a lint violation (indexing into a `dynamic`).

Reply:
Sounds good. You might be able to further reduce casting by making the select variant accept an Object? and defer calling toString until you actually need to Uri parse